AsyncDisplayKit/Layout/ASLayout.mm

```diff
@@ -103,53 +127,61 @@ + (instancetype)flattenedLayoutWithLayoutableObject:(id<ASLayoutable>)layoutable
                      constrainedSizeRange:sizeRange
                                      size:size
                                  position:CGPointNull
-                               sublayouts:sublayouts
-                                flattened:YES];
+                               sublayouts:sublayouts];
+}
+
++ (instancetype)layoutWithLayout:(ASLayout *)layout position:(CGPoint)position
+{
+  return [self layoutWithLayoutableObject:layout.layoutableObject
+                     constrainedSizeRange:layout.constrainedSizeRange
+                                     size:layout.size
+                                 position:position
+                               sublayouts:layout.sublayouts];
 }
 
-- (ASLayout *)flattenedLayoutUsingPredicateBlock:(BOOL (^)(ASLayout *))predicateBlock
+#pragma mark - Layout Flattening
+
+- (ASLayout *)filteredNodeLayoutTree
 {
   NSMutableArray *flattenedSublayouts = [NSMutableArray array];
   
   struct Context {
     ASLayout *layout;
     CGPoint absolutePosition;
-    BOOL visited;
-    BOOL flattened;
   };
   
   // Queue used to keep track of sublayouts while traversing this layout in a BFS fashion.
   std::queue<Context> queue;
-  queue.push({self, CGPointMake(0, 0), NO, NO});
+  queue.push({self, CGPointMake(0, 0)});
   
   while (!queue.empty()) {
-    Context &context = queue.front();
-    if (context.visited) {
-      queue.pop();
-    } else {
-      context.visited = YES;
-      
-      if (predicateBlock(context.layout)) {
-        [flattenedSublayouts addObject:[ASLayout layoutWithLayoutableObject:context.layout.layoutableObject
-                                                       constrainedSizeRange:context.layout.constrainedSizeRange
-                                                                       size:context.layout.size
-                                                                   position:context.absolutePosition
-                                                                 sublayouts:nil
-                                                                  flattened:context.flattened]];
-      }
-      
-      for (ASLayout *sublayout in context.layout.sublayouts) {
-        // Mark layout trees that have already been flattened for future identification of immediate sublayouts
-        BOOL flattened = context.flattened ? : context.layout.flattened;
-        queue.push({sublayout, context.absolutePosition + sublayout.position, NO, flattened});
+    Context context = queue.front();
+    queue.pop();
+
+    if (self != context.layout && context.layout.type == ASLayoutableTypeDisplayNode) {
```

Comment:
This is the main reason I added `layoutableType` — namely, to eliminate the need to class identity check.